scripts/proto/lbann_proto.py

```diff
@@ -371,34 +372,37 @@ def proto(self, model):
     _create_proto_callback(field.message_type.name, field.name)
 _add_to_module_namespace(_proto_callbacks)
 
-def traverse_model(first_module, func):
-    """Traverse a model and apply func to each layer.
+def traverse_layer_graph(start_layers):
+    """Traverse a layer graph.
 
-    first_module is the first module in the model (i.e. the input layer).
-    The model is traversed in a depth-first order except that no layer is
-    visited until all its ancestors have been visited.
+    The traversal starts from the entries in 'start_layers'. The
+    traversal is in depth-first order, except that no layer is visited
+    until all its inputs have been visited.
 
     """
+    layers = []
     visited = set()
-    stack = [first_module]
+    stack = (start_layers
+             if isinstance(start_layers, (list,))
+             else [start_layers])
     while stack:
         l = stack.pop()
+        layers.append(l)
         visited.add(l)
-        func(l)
-        for child in l.children:
-            if (child not in visited
-                and all([parent in visited for parent in child.parents])):
-                stack.append(child)
+        for output in l.outputs:
+            if ((output not in visited)
+                and all([(i in visited) for i in output.inputs])):
+                stack.append(output)
+    return layers
 
-def save_model(filename, first_module, data_layout, mini_batch_size, epochs,
+def save_model(filename, layers, mini_batch_size, epochs,
```

Comment:
1. Let's keep the data layout argument, just make it default to `data_parallel` (as above).
2. I think `first_module` (or `first_layer`, or maybe `first_modules`/`layers` since we could have multiple input layers, I guess) is a better argument name than `layers`, which reads like you need to provide _all_ the layers in the network.

Reply:
1. As far as I'm aware of, the `data_layout` field in the `Model` protobuf message doesn't do anything.
2. I think it might be simpler if we expected the user to provide all the layers. It's fairly straightforward with the `traverse_layer_graph` function.